File: backend/src/main/java/com/apurv/workspace/service/WorkspaceService.java
package com.apurv.workspace.service;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.apurv.workspace.entity.WorkspaceRole;
import com.apurv.auth.entity.User;
import com.apurv.auth.repository.UserRepository;
import com.apurv.common.exception.DuplicateResourceException;
import com.apurv.common.exception.ResourceNotFoundException;
import com.apurv.workspace.dto.InviteMemberRequest;
import com.apurv.workspace.dto.MemberResponse;
import com.apurv.workspace.dto.WorkspaceRequest;
import com.apurv.workspace.dto.WorkspaceResponse;
import com.apurv.workspace.entity.Workspace;
import com.apurv.workspace.entity.WorkspaceMember;
import com.apurv.workspace.repository.WorkspaceMemberRepository;
import com.apurv.workspace.repository.WorkspaceRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class WorkspaceService {

    private final WorkspaceRepository workspaceRepository;
    private final WorkspaceMemberRepository workspaceMemberRepository;
    private final UserRepository userRepository;
    private final WorkspaceAuthorizationService workspaceAuthorizationService;

    public WorkspaceResponse createWorkspace(WorkspaceRequest request, User currentUser) {
        Workspace workspace = new Workspace();
        workspace.setName(request.getName());
        workspace.setDescription(request.getDescription());
        workspace.setOwner(currentUser);
        Workspace savedWorkspace = workspaceRepository.save(workspace);

        WorkspaceMember workspaceMember = new WorkspaceMember();
        workspaceMember.setWorkspace(savedWorkspace);
        workspaceMember.setUser(currentUser);
        workspaceMember.setRole(WorkspaceRole.ADMIN);
        workspaceMemberRepository.save(workspaceMember);

        log.info("Workspace with id {} created by user {}", savedWorkspace.getId(), currentUser.getId());
        return toWorkspaceResponse(savedWorkspace, 1);
    }

    @Transactional(readOnly = true)
    public List<WorkspaceResponse> getAllWorkspaces(User currentUser) {
        List<Workspace> workspaces = workspaceRepository.findAllByMembersId(currentUser.getId());
        return workspaces.stream().map(workspace -> {
            int memberCount = workspaceMemberRepository.countByWorkspaceId(workspace.getId());
            return toWorkspaceResponse(workspace, memberCount);
        }).toList();
    }

    @Transactional(readOnly = true)
    public WorkspaceResponse getWorkspaceById(UUID workspaceId, User currentUser) {
        Workspace workspace = getWorkspace(workspaceId);

        workspaceAuthorizationService.requireRole(workspaceId, currentUser.getId(), WorkspaceRole.ADMIN,
                WorkspaceRole.MEMBER, WorkspaceRole.VIEWER);

        int memberCount = workspaceMemberRepository.countByWorkspaceId(workspace.getId());
        return toWorkspaceResponse(workspace, memberCount);
    }

    public WorkspaceResponse updateWorkspace(UUID workspaceId, WorkspaceRequest request, User currentUser) {
        Workspace workspace = getWorkspace(workspaceId);

        workspaceAuthorizationService.requireRole(workspaceId, currentUser.getId(), WorkspaceRole.ADMIN);

        workspace.setName(request.getName());
        workspace.setDescription(request.getDescription());
        Workspace updatedWorkspace = workspaceRepository.save(workspace);

        log.info("Workspace with id {} updated by user {}", workspaceId, currentUser.getId());
        int memberCount = workspaceMemberRepository.countByWorkspaceId(updatedWorkspace.getId());
        return toWorkspaceResponse(updatedWorkspace, memberCount);
    }

    public void deleteWorkspace(UUID workspaceId, User currentUser) {
        Workspace workspace = getWorkspace(workspaceId);

        if (!workspace.getOwner().getId().equals(currentUser.getId())) {
            throw new AccessDeniedException("Only the workspace owner can delete the workspace");
        }

        workspaceRepository.delete(workspace);
        log.info("Workspace with id {} deleted by user {}", workspaceId, currentUser.getId());
    }

    public MemberResponse inviteMember(UUID workspaceId, InviteMemberRequest request, User currentUser) {
        Workspace workspace = getWorkspace(workspaceId);

        workspaceAuthorizationService.requireRole(workspaceId, currentUser.getId(), WorkspaceRole.ADMIN);

        User invitedUser = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new ResourceNotFoundException("User not found with email: " + request.getEmail()));

        if (workspaceMemberRepository.existsByWorkspaceIdAndUserId(workspaceId, invitedUser.getId())) {
            throw new DuplicateResourceException("User is already a member of the workspace");
        }

        WorkspaceMember workspaceMember = new WorkspaceMember();
        workspaceMember.setWorkspace(workspace);
        workspaceMember.setUser(invitedUser);
        workspaceMember.setRole(request.getRole());
        workspaceMember.setInvitedBy(currentUser);
        WorkspaceMember savedWorkspaceMember = workspaceMemberRepository.save(workspaceMember);

        log.info("User with id {} invited to workspace {} by user {}", invitedUser.getId(), workspaceId,
                currentUser.getId());
        return toMemberResponse(savedWorkspaceMember);
    }

    public void removeMember(UUID workspaceId, UUID userId, User currentUser) {
        Workspace workspace = getWorkspace(workspaceId);

        if (workspace.getOwner().getId().equals(userId)) {
            throw new IllegalArgumentException("Cannot remove the workspace owner from the workspace");
        }

        workspaceAuthorizationService.requireRole(workspaceId, currentUser.getId(), WorkspaceRole.ADMIN);

        WorkspaceMember workspaceMember = workspaceMemberRepository.findByWorkspaceIdAndUserId(workspaceId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Member not found in the workspace"));

        workspaceMemberRepository.delete(workspaceMember);
        log.info("User with id {} removed from workspace {} by user {}", userId, workspaceId, currentUser.getId());
    }

    @Transactional(readOnly = true)
    public List<MemberResponse> getMembers(UUID workspaceId, User currentUser) {
        getWorkspace(workspaceId);

        workspaceAuthorizationService.requireRole(workspaceId, currentUser.getId(), WorkspaceRole.ADMIN);

        List<WorkspaceMember> workspaceMembers = workspaceMemberRepository.findAllByWorkspaceId(workspaceId);
        return workspaceMembers.stream()
                .map(this::toMemberResponse)
                .collect(Collectors.toList());
    }

    private WorkspaceResponse toWorkspaceResponse(Workspace workspace, int memberCount) {
        return WorkspaceResponse.builder()
                .id(workspace.getId())
                .name(workspace.getName())
                .description(workspace.getDescription())
                .ownerId(workspace.getOwner().getId())
                .ownerUsername(workspace.getOwner().getUsername())
                .memberCount(memberCount)
                .createdAt(workspace.getCreatedAt())
                .updatedAt(workspace.getUpdatedAt())
                .build();
    }

    private MemberResponse toMemberResponse(WorkspaceMember workspaceMember) {
        return MemberResponse.builder()
                .userId(workspaceMember.getUser().getId())
                .username(workspaceMember.getUser().getUsername())
                .email(workspaceMember.getUser().getEmail())
                .role(workspaceMember.getRole())
                .joinedAt(workspaceMember.getJoinedAt())
                .build();
    }

    private Workspace getWorkspace(UUID id) {
        return workspaceRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Workspace not found"));
    }
}

File: frontend/src/workspace/service/WorkspaceService.ts
import axios from "axios";
import api from "../../config/axios";
import type ApiResponse from "../../common/types/ApiResponse";
import type WorkspaceResponse from "../types/response/WorkspaceResponse";
import type WorkspaceRequest from "../types/request/WorkspaceRequest";
import type MemberResponse from "../types/response/MemberResponse";
import type InviteMemberRequest from "../types/request/InviteMemberRequest";

function extractError(error: unknown, fallback: string): string {
  if (axios.isAxiosError<ApiResponse<unknown>>(error)) {
    return error.response?.data?.message || error.message || fallback;
  }
  if (error instanceof Error) return error.message;
  return fallback;
}

export async function fetchWorkspaces() {
  try {
    const res = await api.get<ApiResponse<WorkspaceResponse[]>>("/workspaces");
    if (res.data.success && res.data.data) {
      return { success: true as const, data: res.data.data };
    }
    return {
      success: false as const,
      error: res.data.message || "Failed to fetch workspaces.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Failed to fetch workspaces."),
    };
  }
}

export async function createWorkspace(request: WorkspaceRequest) {
  try {
    const res = await api.post<ApiResponse<WorkspaceResponse>>(
      "/workspaces",
      request,
    );
    if (res.data.success && res.data.data) {
      return { success: true as const, data: res.data.data };
    }
    return {
      success: false as const,
      error: res.data.message || "Failed to create workspace.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Failed to create workspace."),
    };
  }
}

export async function updateWorkspace(
  workspaceId: string,
  request: WorkspaceRequest,
) {
  try {
    const res = await api.put<ApiResponse<WorkspaceResponse>>(
      `/workspaces/${workspaceId}`,
      request,
    );
    if (res.data.success && res.data.data) {
      return { success: true as const, data: res.data.data };
    }
    return {
      success: false as const,
      error: res.data.message || "Failed to update workspace.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Failed to update workspace."),
    };
  }
}

export async function deleteWorkspace(workspaceId: string) {
  try {
    const res = await api.delete<ApiResponse<void>>(
      `/workspaces/${workspaceId}`,
    );
    if (res.data.success) {
      return { success: true as const };
    }
    return {
      success: false as const,
      error: res.data.message || "Failed to delete workspace.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Failed to delete workspace."),
    };
  }
}

export async function fetchMembers(workspaceId: string) {
  try {
    const res = await api.get<ApiResponse<MemberResponse[]>>(
      `/workspaces/${workspaceId}/members`,
    );
    if (res.data.success && res.data.data) {
      return { success: true as const, data: res.data.data };
    }
    return {
      success: false as const,
      error: res.data.message || "Failed to fetch members.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Failed to fetch members."),
    };
  }
}

export async function inviteMember(
  workspaceId: string,
  request: InviteMemberRequest,
) {
  try {
    const res = await api.post<ApiResponse<MemberResponse>>(
      `/workspaces/${workspaceId}/members`,
      request,
    );
    if (res.data.success && res.data.data) {
      return { success: true as const, data: res.data.data };
    }
    return {
      success: false as const,
      error: res.data.message || "Failed to invite member.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Failed to invite member."),
    };
  }
}

export async function removeMember(workspaceId: string, userId: string) {
  try {
    const res = await api.delete<ApiResponse<void>>(
      `/workspaces/${workspaceId}/members/${userId}`,
    );
    if (res.data.success) {
      return { success: true as const };
    }
    return {
      success: false as const,
      error: res.data.message || "Failed to remove member.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Failed to remove member."),
    };
  }
}

export async function leaveWorkspace(workspaceId: string) {
  try {
    const res = await api.delete<ApiResponse<void>>(
      `/workspaces/${workspaceId}/members/leave`,
    );
    if (res.data.success) {
      return { success: true as const };
    }
    return {
      success: false as const,
      error: res.data.message || "Failed to leave workspace.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Failed to leave workspace."),
    };
  }
}

const WorkspaceService = {
  fetchWorkspaces,
  createWorkspace,
  updateWorkspace,
  deleteWorkspace,
  fetchMembers,
  inviteMember,
  removeMember,
  leaveWorkspace,
};

export default WorkspaceService;
